odoo/addons/base/models/ir_cron.py

```diff
@@ -291,82 +306,147 @@ def _acquire_one_job(cls, cr, job_ids):
-            ir_cron = env[cls._name]
-
-            # Use the user's timezone to compare and compute datetimes,
-            # otherwise unexpected results may appear. For instance, adding
-            # 1 month in UTC to July 1st at midnight in GMT+2 gives July 30
-            # instead of August 1st!
-            now = fields.Datetime.context_timestamp(ir_cron, datetime.utcnow())
-            past_nextcall = fields.Datetime.context_timestamp(
-                ir_cron, fields.Datetime.to_datetime(job['nextcall']))
-
-            # Compute how many call were missed
-            missed_call = past_nextcall
-            missed_call_count = 0
-            while missed_call <= now:
-                missed_call += interval
-                missed_call_count += 1
-            future_nextcall = missed_call
-
-            # Compute how many time we should run the cron
-            effective_call_count = (
-                     1 if not missed_call_count                    # run at least once
-                else 1 if not job['doall']                         # run once for all
-                else missed_call_count if job['numbercall'] == -1  # run them all
-                else min(missed_call_count, job['numbercall'])     # run maximum numbercall times
-            )
-            call_count_left = max(job['numbercall'] - effective_call_count, -1)
+        if completed:
+            ir_cron._mark_as_done(cron_cr, job)
+        elif os.getenv('ODOO_NOTIFY_CRON_CHANGES'): # See: `_notifydb`
+            ir_cron._cr.postcommit.add(ir_cron._notifydb)
 
-            # The actual cron execution
-            for call in range(effective_call_count):
-                ir_cron._callback(job['cron_name'], job['ir_actions_server_id'], job['id'])
+        cron_cr.commit()
 
-        # Update the cron with the information computed above
-        cron_cr.execute("""
+    @classmethod
+    def _execute_job(cls, job, timed_out_counter=0):
+        """ Execute the job's server action in a loop until it completes.
+        (i.e. the server action reports no remaining records to process)
+        or it fails without any record processed (i.e. the server action crashes and
+        reports no record done)
+        Each call to the ir_action has its own transaction and ir_cron_progress.
+        """
+        with cls.pool.cursor() as job_cr:
+            env = api.Environment(job_cr, job['user_id'], {'lastcall': job['lastcall'], 'cron_id': job['id']})
+            ir_cron = env[cls._name].browse(job['id'])
+            completed = True
+            for i in range(MAX_BATCH_PER_CRON_JOB):
+                ir_cron = ir_cron._add_progress(timed_out=timed_out_counter+1)
+                progress = ir_cron._get_progress()
+                job_cr.commit()
+
+                try:
+                    (done, remaining) = ir_cron._callback(
+                        job['cron_name'], job['ir_actions_server_id'], job['id']
+                    )
```

Comment:
If `ir_cron` is the record, then the parameter `job['id']` is simply useless.

Reply:
I wonder who is to blame for that API :hand_over_mouth:  :stuck_out_tongue_winking_eye: https://github.com/odoo/odoo/blame/870974effeb77f1125394816aa10b96f7f1b1941/odoo/addons/base/ir/ir_cron.py#L93-L94
While I'm +1 to change a few variable names here and there while moving the code. I don't think it is @fdamhaut's duty to cope with our dirty legacy code. I even think we would gain by making the diff not larger than it already is.